### DISCUSSION TITLE
Good practice to use ExternalTaskSensor instead of "depends_on_past"?

### DISCUSSION THREAD

**Discussion by notatallshaw-gts** [apache/airflow#28364] [Q&A]
A colleague ran in to some edge cases where `depends_on_past` was caused issues, in particular running manual DAGs it didn't behave how they expected, nor when the previous task was skipped (due to certain holidays being skipped).

It occurred to me, why not use ExternalTaskSensor instead? Use the dag_id and task_id that points to the task in question but a `execution_date_fn` that calculates the previous business day. This allows much better control over the logic of "when" something should be previously waited upon.

Does this make sense? Is anyone else doing this?

**Comment by potiuk** [comment]
It is quite brittle and makes your DAG contain logic in several places (I guess you have a branch operator that decides to skip execution when there are business holidays, and you would have to add the same logic in the External Task Sensor calculation of the date to check.

Much better and consistent solution to that would be to use a Custom Timetable to schedule your DAG only on business days. This could then be applied accross the board - for all the DAGs that should not run during the business days.

**Reply by notatallshaw-gts** [reply]
> Much better and consistent solution to that would be to use a Custom Timetable to schedule your DAG only on business days. This could then be applied accross the board - for all the DAGs that should not run during the business days.

I agree this might be a good solution for some use cases. Unfortunately for our use case some tasks in the DAG run every week day where as others run every business day.

And if we split them up in to separate DAGs we would need extensive use of `ExternalTaskSensor` with again a fairly complicated `execution_date_fn`.

**Reply by potiuk** [reply]
Ok. In this case, yeah. Separating the logic for business days to a common util and using them in a few places might be the way to go. 

In the future it might be a good input when we implement something that is so far only hinted in [AIP-48](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-48+Data+Dependency+Management+and+Data+Driven+Scheduling)

> Schedule on Dataset change and timetable
> It would be possible that someone might want a DAG to run both on a regular schedule, and whenever an input dataset changes. This may or may not be in this AIP depending on how difficult it is to implement

Having dataset depenency between tasks and schedule in some dags (and combining them - i.e only run when both schedule is met and dataset is produced seems like good fit for your case. I will keep that in mind when this is being worked on as this seems like an interesting case.

**Reply by notatallshaw-gts** [reply]
Thanks for the info, I haven't played around with dataset driven DAGs yet so I don't have good feel for where they are useful, but I will at some point in 2023.
